google/datalab/contrib/bigquery/operators/_bq_extract_operator.py

```diff
@@ -20,23 +20,36 @@ class ExtractOperator(BaseOperator):
   template_fields = ('_table', '_path')
 
   @apply_defaults
-  def __init__(self, table, path, format='csv', csv_options=None, *args, **kwargs):
+  def __init__(self, path, table=None, format='csv', csv_options=None, *args, **kwargs):
     super(ExtractOperator, self).__init__(*args, **kwargs)
     self._table = table
     self._path = path
     self._format = format
     self._csv_options = csv_options or {}
 
   def execute(self, context):
-    if self._table:
-      pydatalab_context = google.datalab.Context.default()
-      source_table = google.datalab.bigquery.Table(self._table, context=pydatalab_context)
-      job = source_table.extract(
-        self._path, format='CSV' if self._format == 'csv' else 'NEWLINE_DELIMITED_JSON',
-        csv_delimiter=self._csv_options.get('delimiter'),
-        csv_header=self._csv_options.get('header'), compress=self._csv_options.get('compress'))
-    else:
-      raise Exception('A table is needed to extract')
+    if not self._table:
+      task_instance = context['task_instance']
+      # If the table is not specified, we fetch it from the output of the execute task, i.e. the
+      # query results table. This could either be a permanent table or a temporary table. If we're
+      # here, it is most likely a temporary table. If it was a permanent one, it would have been
+      # passed in as a param and we wouldn't be here.
+      # TODO(rajivpb): Assert that if we're here, then the table is a temporary one.
+      # TODO(rajivpb):
+      # The task id of the execute task is created by concatenating 'bq_pipeline_execute_task'
+      # and '_id'. This is currently being hard-coded, but consider making this a parameter.
+      # It could require substantial changes to the underlying object model of Pipeline.
+      self._table = task_instance.xcom_pull(task_ids='bq_pipeline_execute_task_id')
```

Comment:
We should have a consistent model for all operators to produce a dictionary of outputs, so as to support operators that have 0, 1, N outputs, as well as evolve over time to introduce additional outputs.

So it would be good if the execute operator produced an output { 'table': table-name }.


Similarly, I'd expect load and extract to also produce results.

Reply:
Done for all operators now.